File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/DisposeOnVanish.cpp
#include <gtest/gtest.h>
#include <Altseed.h>
#include "../EngineTest.h"
#include <iostream>

using namespace std;
using namespace asd;

class ObjectSystem_DisposeOnVanish : public EngineTest
{
private:
	class MessageLayer2D : public Layer2D
	{
	protected:
		void OnDispose()
		{
			wprintf(L"Layer2D(%s):Dispose\n", GetName().c_str());
		}
	};

	class MessageObject2D : public TextureObject2D
	{
	public:
		astring m_name;

	protected:
		void OnDispose()
		{
			wprintf(L"Object2D(%s):Dispose\n", m_name.c_str());
		}
	};

	int time;
	shared_ptr<MessageLayer2D> layer1;
	shared_ptr<MessageObject2D> object1;

public:
	ObjectSystem_DisposeOnVanish(bool isOpenGLMode)
		: EngineTest(asd::ToAString("DisposeOnVanish"), isOpenGLMode, 120)
		, time(0)
	{
	}

protected:
	void OnStart()
	{
		auto scene = make_shared<Scene>();
		layer1 = make_shared<MessageLayer2D>();
		auto layer2 = make_shared<MessageLayer2D>();
		object1 = make_shared<MessageObject2D>();
		auto object2 = make_shared<MessageObject2D>();

		Engine::ChangeScene(scene);
		scene->AddLayer(layer1);
		scene->AddLayer(layer2);
		layer1->AddObject(object1);
		layer1->AddObject(object2);

		layer1->SetName(ToAString("Layer1"));
		layer2->SetName(ToAString("Layer2"));
		object1->m_name = ToAString("Object1");
		object2->m_name = ToAString("Object2");
	}

	void OnUpdated()
	{
		time++;
		if (time == 10)
		{
			printf("Vanish Object1\n");
			object1->Vanish();
		}
		if (time == 60)
		{
			printf("Vanish Layer1\n");
			layer1->Vanish();
		}
	}

	void OnFinish()
	{
		printf("Terminate\n");
	}
};

ENGINE_TEST(ObjectSystem, DisposeOnVanish)

File: Dev/unitTest_Engine_cs/ObjectSystem2D/DisposeOnVanish.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.ObjectSystem2D
{
    class DisposeOnVanish : EngineTest
    {
        class MessageLayer : Layer2D
        {
            protected override void OnDispose()
            {
                Console.WriteLine("Layer2D({0}):Dispose", Name);
            }
        }

        class MessageObject : TextureObject2D
        {
            private string name { get; set; }

            public MessageObject(string name)
            {
                this.name = name;
            }

            protected override void OnDispose()
            {
                Console.WriteLine("Object2D({0}):Dispose", name);
            }
        }

        int time = 0;
        Layer2D layer1;
        TextureObject2D obj1;

        public DisposeOnVanish() : base(120)
        {
        }

        protected override void OnStart()
        {
            var scene = new Scene();
            layer1 = new MessageLayer() { Name = "Layer1" };
            var layer2 = new MessageLayer() { Name = "Layer2" };
            obj1 = new MessageObject("Object1");
            var obj2 = new MessageObject("Object2");

            Engine.ChangeScene(scene);
            scene.AddLayer(layer1);
            scene.AddLayer(layer2);
            layer1.AddObject(obj1);
            layer1.AddObject(obj2);
        }

        protected override void OnUpdated()
        {
            time++;
            if(time == 10)
            {
                Console.WriteLine("Vanish Object1");
                obj1.Dispose();
            }
            if(time == 60)
            {
                Console.WriteLine("Vanish Layer1");
                layer1.Dispose();
            }
        }

        protected override void OnFinish()
        {
            Console.WriteLine("Terminate");
        }
    }
}
